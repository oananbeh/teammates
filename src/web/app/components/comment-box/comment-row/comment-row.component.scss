.dot-xs {
  font-size: 0.3em;
}
