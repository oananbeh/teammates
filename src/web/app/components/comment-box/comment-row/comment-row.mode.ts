/**
 * Mode of current comment row.
 */
export enum CommentRowMode {
  /**
   * Add new comment.
   */
  ADD,

  /**
   * Edit existing comment.
   */
  EDIT,
}
