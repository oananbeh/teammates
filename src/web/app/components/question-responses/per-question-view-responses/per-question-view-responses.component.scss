.sortable-header {
  cursor: pointer;

  i {
    margin-left: 0.3em;
  }

  .fa-stack {
    width: 20%;
  }

  button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    text-align: left;
    width: 100%;
    font: inherit;
    cursor: inherit;
  }
}

.color-neutral {
  color: gray;
}

.actions-cell,
.actions-cell-narrow {
  min-width: 160px;
}

@media (width >= 1000px) {
  .actions-cell {
    min-width: 300px;
  }

  .actions-cell-narrow {
    min-width: 160px;
  }
}
