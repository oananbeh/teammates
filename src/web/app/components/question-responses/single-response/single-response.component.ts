import { Component, Input } from '@angular/core';
import {
  FeedbackQuestionDetails,
  FeedbackQuestionType,
  FeedbackResponseDetails,
} from '../../../../types/api-output';

/**
 * The component that will map a generic response to its specialized response view component.
 */
@Component({
  selector: 'tm-single-response',
  templateUrl: './single-response.component.html',
  styleUrls: ['./single-response.component.scss'],
})
export class SingleResponseComponent {

  @Input() responseDetails: FeedbackResponseDetails = {
    questionType: FeedbackQuestionType.TEXT,
  };

  @Input() questionDetails: FeedbackQuestionDetails = {
    questionType: FeedbackQuestionType.TEXT,
    questionText: '',
  };

  @Input() isStudentPage: boolean = false;
  @Input() statistics: string = '';
  @Input() giverEmail: string = '';
  @Input() recipientEmail: string = '';

  // enum
  FeedbackQuestionType: typeof FeedbackQuestionType = FeedbackQuestionType;

}
