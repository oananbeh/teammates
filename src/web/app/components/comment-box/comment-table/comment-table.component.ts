import { Component, EventEmitter, Input, Output } from '@angular/core';
import {
  FeedbackResponseComment, FeedbackVisibilityType, ResponseOutput,
} from '../../../../types/api-output';
import { collapseAnim } from '../../teammates-common/collapse-anim';
import { CommentRowModel } from '../comment-row/comment-row.component';
import { CommentRowMode } from '../comment-row/comment-row.mode';

/**
 * Model for CommentTableComponent.
 */
export interface CommentTableModel {
  commentRows: CommentRowModel[];
  newCommentRow: CommentRowModel;

  isAddingNewComment: boolean;
  isReadOnly: boolean;
}

/**
 * Component for the comments table.
 */
@Component({
  selector: 'tm-comment-table',
  templateUrl: './comment-table.component.html',
  styleUrls: ['./comment-table.component.scss'],
  animations: [collapseAnim],
})
export class CommentTableComponent {

  // enum
  CommentRowMode: typeof CommentRowMode = CommentRowMode;

  @Input()
  response?: ResponseOutput;

  @Input()
  questionShowResponsesTo: FeedbackVisibilityType[] = [];

  @Input()
  displayAddCommentButton: boolean = false;

  @Input()
  model: CommentTableModel = {
    commentRows: [],
    newCommentRow: {
      commentEditFormModel: {
        commentText: '',
        isUsingCustomVisibilities: false,
        showCommentTo: [],
        showGiverNameTo: [],
      },
      isEditing: true,
    },
    isAddingNewComment: true,
    isReadOnly: false,
  };

  @Input()
  shouldHideClosingButtonForNewComment: boolean = false;

  @Output()
  modelChange: EventEmitter<CommentTableModel> = new EventEmitter();

  @Output()
  saveNewCommentEvent: EventEmitter<any> = new EventEmitter();

  @Output()
  deleteCommentEvent: EventEmitter<number> = new EventEmitter();

  @Output()
  updateCommentEvent: EventEmitter<number> = new EventEmitter();

  /**
   * Triggers the delete comment event.
   */
  triggerDeleteCommentEvent(index: number): void {
    this.deleteCommentEvent.emit(index);
  }

  /**
   * Triggers the update comment event.
   */
  triggerUpdateCommentEvent(index: number): void {
    this.updateCommentEvent.emit(index);
  }

  /**
   * Triggers the add new comment event.
   */
  triggerSaveNewCommentEvent(): void {
    this.saveNewCommentEvent.emit();
  }

  /**
   * Triggers the change of comment rows for the form.
   */
  triggerCommentRowChange(index: number, data: CommentRowModel): void {
    const newCommentRows: CommentRowModel[] = JSON.parse(JSON.stringify(this.model.commentRows));
    newCommentRows[index] = data;
    this.triggerModelChange('commentRows', newCommentRows);
  }

  /**
   * Triggers the change of the model for the form.
   */
  triggerModelChange(field: string, data: any): void {
    this.modelChange.emit({ ...this.model, [field]: data });
  }

  /**
   * Tracks by index.
   */
  trackByIndex(index: number): string {
    return index.toString();
  }

  /**
   * Handles the close editing event.
   */
  handleCloseEditingCommentRowEvent(index: number): void {
    const newRowModel: CommentRowModel = JSON.parse(JSON.stringify(this.model.commentRows[index]));
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const originalComment: FeedbackResponseComment = newRowModel.originalComment!;
    newRowModel.commentEditFormModel = {
      commentText: originalComment.commentText,

      isUsingCustomVisibilities: false,
      showCommentTo: originalComment.showCommentTo,
      showGiverNameTo: originalComment.showGiverNameTo,
    };
    newRowModel.isEditing = false;
    this.triggerCommentRowChange(index, newRowModel);
  }

  /**
   * Handles adding new comment button click event.
   */
  handleAddingNewCommentEvent(): void {
    this.triggerModelChange('isAddingNewComment', true);
  }
}
