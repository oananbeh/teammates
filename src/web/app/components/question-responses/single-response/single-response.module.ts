import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { SingleResponseComponent } from './single-response.component';
import { QuestionResponseModule } from '../../question-types/question-response/question-response.module';

/**
 * Module for single response component.
 */
@NgModule({
  declarations: [
    SingleResponseComponent,
  ],
  exports: [
    SingleResponseComponent,
  ],
  imports: [
    CommonModule,
    QuestionResponseModule,
  ],
})
export class SingleResponseModule { }
