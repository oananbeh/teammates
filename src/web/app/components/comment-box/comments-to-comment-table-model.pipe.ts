import { Pipe, PipeTransform } from '@angular/core';
import { CommentTableModel } from './comment-table/comment-table.component';
import { CommentToCommentRowModelPipe } from './comment-to-comment-row-model.pipe';
import { CommentOutput } from '../../../types/api-output';

/**
 * Transforms comments to readonly comment table model.
 */
@Pipe({
  name: 'commentsToCommentTableModel',
})
export class CommentsToCommentTableModelPipe implements PipeTransform {
  constructor(private commentToCommentRowModel: CommentToCommentRowModelPipe) {
  }
  transform(comments: CommentOutput[], isReadOnly: boolean, timezone?: string): CommentTableModel {
    return {
      isReadOnly,
      commentRows: comments.map((comment: CommentOutput) => {
        return this.commentToCommentRowModel.transform(comment, timezone);
      }),
      newCommentRow: {
        commentEditFormModel: {
          commentText: '',
          isUsingCustomVisibilities: false,
          showCommentTo: [],
          showGiverNameTo: [],
        },
        isEditing: false,
      },
      isAddingNewComment: false,
    };
  }

}
