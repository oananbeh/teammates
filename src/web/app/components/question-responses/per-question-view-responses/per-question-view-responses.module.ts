import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';

import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { PerQuestionViewResponsesComponent } from './per-question-view-responses.component';
import {
  ResponseModerationButtonModule,
} from '../../../pages-instructor/instructor-session-result-page/response-moderation-button/response-moderation-button.module';
import { CommentBoxModule } from '../../comment-box/comment-box.module';
import { TeammatesCommonModule } from '../../teammates-common/teammates-common.module';
import { SingleResponseModule } from '../single-response/single-response.module';

/**
 * Module for component to display list of responses for one question.
 */
@NgModule({
  declarations: [
    PerQuestionViewResponsesComponent,
  ],
  exports: [
    PerQuestionViewResponsesComponent,
  ],
  providers: [
    NgbActiveModal,
  ],
  imports: [
    CommentBoxModule,
    CommonModule,
    TeammatesCommonModule,
    RouterModule,
    SingleResponseModule,
    ResponseModerationButtonModule,
  ],
})
export class PerQuestionViewResponsesModule { }
