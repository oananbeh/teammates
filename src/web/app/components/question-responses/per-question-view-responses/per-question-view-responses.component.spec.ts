import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { RouterModule } from '@angular/router';
import {
  CommentOutput, FeedbackQuestion, FeedbackTextQuestionDetails,
  FeedbackTextResponseDetails, ResponseOutput,
} from 'src/web/types/api-output';
import SpyInstance = jest.SpyInstance;
import { PerQuestionViewResponsesComponent } from './per-question-view-responses.component';
import { FeedbackResponsesService } from '../../../../services/feedback-responses.service';
import {
  CommentVisibilityType, FeedbackParticipantType, FeedbackQuestionType,
  NumberOfEntitiesToGiveFeedbackToSetting,
} from '../../../../types/api-request';
import {
  ResponseModerationButtonModule,
} from '../../../pages-instructor/instructor-session-result-page/response-moderation-button/response-moderation-button.module';
import { CommentBoxModule } from '../../comment-box/comment-box.module';
import { CommentRowModel } from '../../comment-box/comment-row/comment-row.component';
import { CommentTableModel } from '../../comment-box/comment-table/comment-table.component';
import { RichTextEditorModule } from '../../rich-text-editor/rich-text-editor.module';
import { TeammatesCommonModule } from '../../teammates-common/teammates-common.module';
import { SingleResponseModule } from '../single-response/single-response.module';

describe('PerQuestionViewResponsesComponent', () => {
  let component: PerQuestionViewResponsesComponent;
  let fixture: ComponentFixture<PerQuestionViewResponsesComponent>;

  let feedbackResponsesService: FeedbackResponsesService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [PerQuestionViewResponsesComponent],
      imports: [
        SingleResponseModule,
        CommentBoxModule,
        TeammatesCommonModule,
        HttpClientTestingModule,
        RichTextEditorModule,
        RouterModule,
        ResponseModerationButtonModule,
      ],
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(PerQuestionViewResponsesComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    feedbackResponsesService = TestBed.inject(FeedbackResponsesService);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  const commentOutput: CommentOutput = {
    commentGiverName: 'Jennie Kim',
    lastEditorName: 'Jennie Kim',
    commentGiver: 'Jennie Kim',
    lastEditorEmail: 'jenniekim@gmail.com',
    feedbackResponseCommentId: 3,
    commentText: 'commentText',
    createdAt: 0,
    lastEditedAt: 0,
    isVisibilityFollowingFeedbackQuestion: false,
    showGiverNameTo: [],
    showCommentTo: [],
  };

  const responseOutput: ResponseOutput = {
    isMissingResponse: false,
    responseId: 'resp-id-101',
    giver: 'Jennie Kim',
    giverEmail: 'jenniekim@gmail.com',
    giverTeam: 'Tutorial Group 135',
    giverSection: 'section2',
    recipient: 'Lisa Mano',
    recipientTeam: 'Tutorial Group 246',
    recipientEmail: 'lisamano@gmail.com',
    recipientSection: 'section2',
    responseDetails: {
      answer: '<p>Lisa is a good classmate </p>',
    } as FeedbackTextResponseDetails,
    instructorComments: [],
    participantComment: commentOutput,
  };

  const feedbackQuestion: FeedbackQuestion = {
    feedbackQuestionId: 'feedbackQuestion22',
    questionNumber: 22,
    questionBrief: 'What did you think of the class contribution of this classmate?',
    questionDescription: '',
    questionDetails: {
      shouldAllowRichText: true,
    } as FeedbackTextQuestionDetails,
    questionType: FeedbackQuestionType.TEXT,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.STUDENTS_EXCLUDING_SELF,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    showResponsesTo: [],
    showGiverNameTo: [],
    showRecipientNameTo: [],
    customNumberOfEntitiesToGiveFeedbackTo: 0,
  };

  const commentRowModel: CommentRowModel = {
    commentEditFormModel: {
      commentText: '',
      isUsingCustomVisibilities: false,
      showCommentTo: [CommentVisibilityType.RECIPIENT],
      showGiverNameTo: [CommentVisibilityType.RECIPIENT],
    },
    isEditing: false,
  };

  const commentTableModel: CommentTableModel = {
    commentRows: [],
    newCommentRow: commentRowModel,
    isAddingNewComment: false,
    isReadOnly: false,
  };

  const instructorCommentTableModel: Record<string, CommentTableModel> = {
    'resp-id-101': commentTableModel,
  };

  it('should snap response with comments', () => {
    component.question = feedbackQuestion;
    component.instructorCommentTableModel = instructorCommentTableModel;
    component.responses = [responseOutput];
    const feedbackResponseSpy: SpyInstance = jest.spyOn(feedbackResponsesService,
      'isFeedbackResponsesDisplayedOnSection')
      .mockReturnValue(true);

    component.ngOnInit();
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();

    expect(feedbackResponseSpy).toHaveBeenCalledTimes(1);
    expect(JSON.stringify(component.responsesToShow[0]))
      .toBe(JSON.stringify(responseOutput));
  });
});
