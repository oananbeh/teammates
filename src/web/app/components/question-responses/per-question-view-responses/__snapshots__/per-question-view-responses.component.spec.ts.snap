// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerQuestionViewResponsesComponent should snap response with comments 1`] = `
<tm-per-question-view-responses
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  deleteCommentEvent={[Function EventEmitter_]}
  feedbackResponsesService={[Function FeedbackResponsesService]}
  groupByTeam={[Function Boolean]}
  indicateMissingResponses={[Function Boolean]}
  instructorCommentTableModel={[Function Object]}
  instructorCommentTableModelChange={[Function EventEmitter_]}
  isDisplayOnly="false"
  isExpandAll="false"
  ngbModal={[Function _NgbModal]}
  question={[Function Object]}
  responses={[Function Array]}
  responsesToShow={[Function Array]}
  saveNewCommentEvent={[Function EventEmitter_]}
  section=""
  sectionType={[Function String]}
  session={[Function Object]}
  showGiver={[Function Boolean]}
  showRecipient={[Function Boolean]}
  sortBy="0"
  sortOrder={[Function Number]}
  statistics=""
  tableComparatorService={[Function TableComparatorService]}
  updateCommentEvent={[Function EventEmitter_]}
>
  <div>
    <table
      class="table table-bordered"
      id="response-table"
    >
      <thead>
        <tr
          class="header"
        >
          <th
            aria-sort="none"
            class="sortable-header"
          >
            <button>
               Team 
              <span
                aria-hidden="true"
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </button>
          </th>
          <th
            aria-sort="none"
            class="sortable-header"
          >
            <button>
               Giver 
              <span
                aria-hidden="true"
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </button>
          </th>
          <th
            aria-sort="none"
            class="sortable-header"
          >
            <button>
               Team 
              <span
                aria-hidden="true"
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </button>
          </th>
          <th
            aria-sort="none"
            class="sortable-header"
          >
            <button>
               Recipient 
              <span
                aria-hidden="true"
                class="fa-stack"
              >
                <i
                  class="fas fa-sort"
                />
              </span>
            </button>
          </th>
          <th>
            Response
          </th>
          <th>
            Comment for Response
          </th>
          <th
            class="text-center no-print actions-cell"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Tutorial Group 135
          </td>
          <td>
             Jennie Kim  (jenniekim@gmail.com) 
          </td>
          <td>
             Tutorial Group 246 
          </td>
          <td>
             Lisa Mano  (lisamano@gmail.com) 
          </td>
          <td>
            <tm-single-response>
            </tm-single-response>
          </td>
          <td>
            <div>
              commentText
            </div>
          </td>
          <td
            class="text-center no-print"
          >
            <button
              class="btn btn-light btn-sm btn-margin-left"
              id="btn-add-comment"
            >
              Add Comment 
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</tm-per-question-view-responses>
`;
